src/io.jl

```diff
@@ -223,9 +226,530 @@ function readg(gfile; set_time=nothing)
+        time_index::Int;
+        geqdsk_cocos::Int=0,
+        dd_cocos::Int=11,
+        cocos_clockwise_phi::Bool=false,
+        add_derived::Bool=false,
+    )
+
+Utility for writing equilibrium data from a GEQDSKFile to IMAS DD.
+
+$geqdsk2imas_generic
+
+    g: GEQDSKFile instance
+    dd: IMAS data dictionary
+    time_index: index of the destination for equilibrium data within equilibrium.time_slice
+    geqdsk_cocos: coordinate convention indentifier of the incoming GEQDSK set. 0 to auto detect.
+    dd_cocos: coordinate convention identifier of the DD. Should normally be left at 11.
+    cocos_clockwise_phi: hint for deciding COCOS, because GEQDSK probably doesn't constrain this.
+    add_derived: switch to use functions within EFIT.jl to extend equilibrium data
+"""
+function geqdsk2imas!(
+    g::GEQDSKFile,
+    dd::IMASdd.dd,
+    time_index::Int;
+    geqdsk_cocos::Int=0,
+    dd_cocos::Int=11,
+    cocos_clockwise_phi::Bool=false,
+    add_derived::Bool=false,
+)
+    if ismissing(dd.dataset_description.data_entry, :pulse)
+        dd.dataset_description.data_entry.pulse = parse(Int, split(split(g.file, ".")[end-1], "g")[2])
+    end
+    geqdsk2imas!(
+        g, dd.equilibrium, time_index;
+        wall=dd.wall, geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos, cocos_clockwise_phi=cocos_clockwise_phi,
+        add_derived=add_derived,
+    )
+end
+
+"""
+    function geqdsk2imas!(
+        g::GEQDSKFile,
+        eq::IMASdd.equilibrium,
+        time_index::Int;
+        wall=nothing,
+        geqdsk_cocos::Int=0,
+        dd_cocos::Int=11,
+        cocos_clockwise_phi::Bool=false,
+        add_derived::Bool=false,
+    )
+
+Utility for writing equilibrium data from a GEQDSKFile to IMAS equilibrium IDS
+
+$geqdsk2imas_generic
+
+    g: GEQDSKFile instance
+    eq: Reference to equilibrium IDS
+    time_index: index of the destination for equilibrium data within equilibrium.time_slice
+    wall: Optional reference to wall IDS
+    geqdsk_cocos: coordinate convention indentifier of the incoming GEQDSK set. 0 to auto detect.
+    dd_cocos: coordinate convention identifier of the DD. Should normally be left at 11.
+    cocos_clockwise_phi: hint for deciding COCOS, because GEQDSK probably doesn't constrain this.
+    add_derived: switch to use functions within EFIT.jl to extend equilibrium data
+"""
+function geqdsk2imas!(
+    g::GEQDSKFile,
+    eq::IMASdd.equilibrium,
+    time_index::Int;
+    wall=nothing,
+    geqdsk_cocos::Int=0,
+    dd_cocos::Int=11,
+    cocos_clockwise_phi::Bool=false,
+    add_derived::Bool=false,
+)
+    if geqdsk_cocos == 0
+        geqdsk_cocos = CoordinateConventions.identify_cocos(
+            sign(g.bcentr), sign(g.current), sign(g.qpsi[1]), sign(g.psi[end] - g.psi[1]), cocos_clockwise_phi,
+        )[1]
+        println("  identified COCOS=$geqdsk_cocos")
+    end
+    tc = CoordinateConventions.transform_cocos(geqdsk_cocos, dd_cocos)
+
+    # Make sure top level stuff is defined
```

Comment:
I would suggest leaning on `IMASdd`'s time handling capabilities.  I do need to document these better, so that people are more aware!

A `dd` defines a `global_time` at which all time-dependent operations take place. In your case you should do:
```julia
IMAS.top_dd(eq).global_time = g.time # set the global_time based on the gEQKDS time

resize(eq.time_slice) # this resizes time_slice (if neeed) and sets eq.time_slice[].time (note that in IMAS eq.time is an IMAS expression, but IMASdd does not have expressions)

eqt = eq.time_slice[] # this gives you the time_slice at the current global_time

@ddtime(eq.vacuum_toroidal_field.b0 = g.bcentr .* tc["B"]) # this macro detects that `b0` is a time array and inserts data at the current global_time
```

If you want to be nice you should restore the original `global_time` that was set prior entering your functions.
```julia
dd = IMAS.top_dd(eq)
original_global_time = dd.global_time
try
   dd.global_time = g.time
   ...
finally
   dd.global_time = original_global_time
end
```

Reply:
@orso82 Thanks